### DISCUSSION TITLE
Some tiles are sometimes not rendered

### DISCUSSION THREAD

**Discussion by dennisguse** [mapsforge/mapsforge#1331] [Q&A] [answered]
Some tiles are sometimes not rendered in OSMDashboard (online as well as offline maps). This "feature" is available for quite some time.
We tried:
* multithreading on/off
* online/offline maps
* purging cache (helps immediately, but on rerender(?) tiles are missing again)

More infos: https://github.com/OpenTracksApp/OSMDashboard/issues/160

Now, the question is: how to debug it from here?
We don't see any logcats or other error messages.

**Comment by devemux86** [comment]
Mapsforge samples work properly all this time, so you can compare your code with them.
(it appears that the memory tile cache is not created correctly)

- [GettingStarted](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/GettingStarted.java)
- [mapsforge-samples-android](https://github.com/mapsforge/mapsforge/tree/master/mapsforge-samples-android)

**Reply by dennisguse** [reply]
@pstorch FYI

**Reply by pstorch** [reply]
> (it appears that the memory tile cache is not created correctly)

Yes, that is very likely, but I can't find it yet.

I have based the implementation on the mapsforge android samples. When I compare 

https://github.com/OpenTracksApp/OSMDashboard/blob/2d1dff882e732fac6f2298c99ec097c4e49e1283/src/main/java/de/storchp/opentracks/osmplugin/MapsActivity.java#L268-L270

with

https://github.com/mapsforge/mapsforge/blob/0cba2d25b091c0af8c1a37ea939e7094e14ec76e/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/SamplesBaseActivity.java#L122-L124

it looks very similar.

**Comment by devemux86** [comment]
Check if you changed the `Parameters` class.
Especially `Parameters.SQUARE_FRAME_BUFFER`.

**Reply by pstorch** [reply]
No, I don't change anything the `Parameters` class.

**Comment by ludwigb** [comment]
If not all tiles are rendered, is likely an indication that your TileCache
is not big enough to hold all tiles required.
I had only glimpsed through your code, but I saw that you reimplemented
some of the methods that provide input to the configuration of the
TileCache. This might be a red herring, but worthwhile checking. I would
recommend finding out how large your TileCache actually is and comparing it
against our reference implementation.

**Reply by pstorch** [reply]
Thanks for you help, but I'm still stuck. Can you point me to the differences you saw? When I compare my implementation, it looks the same to the one in the mapsforge SamplesBaseActivity as pointed out in comment https://github.com/mapsforge/mapsforge/discussions/1331#discussioncomment-3086590.

**Reply by pstorch** [reply]
> recommend finding out how large your TileCache actually is

How do I do that?

How do I increase the TileCache properly? Increase the numbers for ScreenRatio, TileSize or OverdrawFactor? What are reasonable values?

**Comment by devemux86** [comment]
The simplest sample is the [GettingStarted](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/GettingStarted.java).

**Reply by pstorch** [reply]
The GettingStarted creates the TilesCache this way:

https://github.com/mapsforge/mapsforge/blob/0cba2d25b091c0af8c1a37ea939e7094e14ec76e/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/GettingStarted.java#L102-L104

Is also similar to my implementation:

https://github.com/OpenTracksApp/OSMDashboard/blob/2d1dff882e732fac6f2298c99ec097c4e49e1283/src/main/java/de/storchp/opentracks/osmplugin/MapsActivity.java#L268-L270

`getPersistableId` is just a name and `getScreenRatio` is `1.0f` also the same value as in the `GettingStarted` sample.

So where do you see the difference? I'm sorry when I bother you. I just don't see it.

**Comment by applantation** [comment]
On Sat, 9 Jul 2022 at 09:02, Peter Storch ***@***.***> wrote:

> recommend finding out how large your TileCache actually is
>
> How do I do that?
>
TileCache.getCapacity()

> How do I increase the TileCache properly? Increase the numbers for
> ScreenRatio, TileSize or OverdrawFactor? What are reasonable values?
>

These all reflect different aspects: if the TileSize increases, you need
more physical space, the ScreenRatio is how much of the screen is covered
by the map (if your map only covers a small part of the screen, a smaller
cache will do), the OverdrawFactor controls how much extra space you
allocate to allow for movement of the map without having to redraw the
tiles.)

I would suggest increasing the OverdrawFactor significantly -- regardless
of this maybe not being the right solution, you are trying to narrow the
problem down. If the problem persists, maybe you have a different problem.
(My hunch would be now that maybe you have not set permissions correctly
and are not using the cache at all. Check for error messages.)

The sample programs and the programs of many developers who use mapsforge
work. So the problem lies on your side somewhere.

**Reply by pstorch** [reply]
Thanks for the insights. I'll check the Capacity.

Actually I've made the `overdrawFactor` configurable between 1 and 2. And @dennisguse reported me that on his device the higher the `overdrawFactor` the more blank tiles he gets. So for now the best result is with `overdrawFactor=1` (the lowest value).

Permissions: that's maybe an interesting point. Does the `TileCache` need external storage permission? This was removed from the app. We only work with SAF. On the other hand we also tried with `persistent=false` for the `TileCache` and this didn't change anything.

**Reply by pstorch** [reply]
> TileCache.getCapacity()

With overdrawFactor=1.2 I get capacity=2000 and capacityFirstLevel=25.

Is that ok or too low?

**Reply by pstorch** [reply]
With overdrawFactor=1.4 I get capacity=2000 and capacityFirstLevel=36.

**Comment by applantation** [comment, marked as answer]
Well, this obviously depends on the device: 2000 tiles should certainly be
enough to cover any device, but 25 tiles for the first level cache is not
much, 4 by 6, so if (as I suspect) you are for some reason (permissions?)
not writing to disk, then you have a problem.
Checked permissions? Checked how many tiles the sample up stores in the
cache? Checked permissions of the sample app? (I am a bit out of the loop
here, the Android permissions have changed a lot and it might even depend
on your min/max Android version in the manifest, IIRC, even the sample app
could not be up to date with latest storage requirements for latest Android
versions.)

**Comment by devemux86** [comment]
Mapsforge Android Samples app is up to date and works fine even on Android 13.
It does not need any storage permission as cache uses the app specific directory.

Better create a minimal app based on [GettingStarted](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/GettingStarted.java)
and study Mapsforge samples to understand how to properly use it.

**Comment by pstorch** [comment]
I just managed to run the Mapsforge Android Samples and the GettingStarted sample show the same TileCache capacity:
```

2022-07-10 11:39:11.690 27458-27458/org.mapsforge.samples.android I/AndroidUtil: TILECACHE INMEMORY SIZE: 25
2022-07-10 11:39:11.692 27458-27458/org.mapsforge.samples.android I/AndroidUtil: TILECACHE FILE SIZE: 2000

```

**Comment by pstorch** [comment]
I've build a version with a double sized TileCache and a user with this blank tile problem reported that they are gone. Even when increasing the overdrawFactor to 2.
So looks like a too small TileCache was the issue.
